integration/connector_test.go

```diff
@@ -0,0 +1,430 @@
+	Value   string                 `json:"value"`
+}
+
+type commandResponsePayload struct {
+	commandPayload
+	Status int `json:"status"`
+}
+
+type ConnectorSuite struct {
+	suite.Suite
+
+	mqttClient  mqtt.Client
+	dittoClient *ditto.Client
+
+	cfg      *testConfig
+	thingCfg *thingConfig
+
+	thingURL   string
+	featureURL string
+}
+
+const (
+	envVariablesPrefix    = "SCT"
+	featureName           = "ConnectorTestFeature"
+	propertyName          = "testProperty"
+	commandName           = "testCommand"
+	commandResponseFormat = "response[%s]"
+)
+
+func (suite *ConnectorSuite) SetupSuite() {
+	cfg := &testConfig{}
+
+	suite.T().Log(getConfigHelp(*cfg, envVariablesPrefix))
+
+	if err := initConfigFromEnv(cfg, envVariablesPrefix); err != nil {
+		suite.T().Skip(err)
+	}
+
+	suite.T().Logf("test config: %+v", *cfg)
+
+	opts := mqtt.NewClientOptions().
+		AddBroker(cfg.Broker).
+		SetClientID(uuid.New().String()).
+		SetKeepAlive(30 * time.Second).
+		SetCleanSession(true).
+		SetAutoReconnect(true)
+
+	mqttClient := mqtt.NewClient(opts)
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		require.NoError(suite.T(), token.Error(), "connect to MQTT broker")
+	}
+
+	thingCfg, err := getThingConfig(mqttClient)
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "get thing config")
+	}
+
+	suite.T().Logf("thing config: %+v", *thingCfg)
+
+	dittoClient, err := ditto.NewClientMQTT(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "initialize ditto client")
+	}
+
+	feature := &model.Feature{}
+	feature.WithProperty(propertyName, "testValue")
+
+	cmd := things.NewCommand(model.NewNamespacedIDFrom(thingCfg.DeviceID)).Twin().Feature(featureName).
+		Modify(feature)
+	msg := cmd.Envelope(protocol.WithResponseRequired(false))
+
+	err = dittoClient.Send(msg)
+	require.NoError(suite.T(), err, "create test feature")
+
+	dittoClient.Subscribe(func(requestID string, msg *protocol.Envelope) {
+		if msg.Path != fmt.Sprintf("/features/%s/inbox/messages/%s", featureName, commandName) {
+			suite.T().Logf("unexpected command: %s", msg.Path)
+			return
+		}
+
+		headers := protocol.NewHeaders(protocol.WithCorrelationID(msg.Headers.CorrelationID()),
+			protocol.WithContentType("application/json"))
+
+		value, ok := msg.Value.(string)
+		if !ok {
+			suite.T().Errorf("unexpected message payload: %v, %T", msg.Value, msg.Value)
+			return
+		}
+
+		response := fmt.Sprintf(commandResponseFormat, value)
+
+		reply := &protocol.Envelope{
+			Topic:   msg.Topic,
+			Headers: headers,
+			Path:    strings.Replace(msg.Path, "/inbox/", "/outbox/", 1),
+			Value:   response,
+			Status:  http.StatusNoContent,
+		}
+
+		if err := dittoClient.Reply(requestID, reply); err != nil {
+			suite.T().Error("failed to send response:", err)
```

Comment:
This call is in goroutine. Calling require.NoError will probably have unexpected consequences, as it invokes TestingT.failNow().

Reply:
Not being able to send a response will ultimately end up in failing the Command test because it expects a response within a certain timeout. So a failure here will also fail the actual test as well, with the code staying as-is. I think we can keep it this way.

Reply:
This is an async callback which will be called outside of our test case methods. If this callback fails in anything it's trying to do, then the actual test expecting the feature to send a response back to the backend will fail (the testCommand test method)

Reply:
Yes, it'll fail but we cannot bypass the test message because the user will not be able to understand the reason behind this fail.

Reply:
I think we can fail the tests asynchronously, I will research to see how we should do it properly.

Reply:
Could you please share the research? We ended with FailNow in the current source code state where we have API doc statement:
```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.
```